### PR TITLE
Use Best Practices and Clean Up the FE Configuration Files (tsconfig, eslintrc, prettierrc, etc.)

### PR DESCRIPTION
**Pull request by vahidbazzaz**
In this PR, we clean up tsconfig, prettierrc, and eslintrc, apply best practices, write down folder structures (at least for components), file names conventions, variable names conventions, type creation/enforcement, where is business logic vs view logic, global variables (apollo reactive variable), and explain what packages in package.json do.

### CODE REVIEW

**Comment by vahidbazzaz** [comment]
[Use Best Practices and Clean Up the FE Configuration Files (tsconfig, eslintrc, prettierrc, etc.)](https://trello.com/c/zMdX4v3R)

File: .eslintrc [hunk 1]
```diff
@@ -9,12 +9,32 @@
   ],
   "plugins": ["jsx-a11y"],
   "rules": {
+
     // driving me nuts and would have involved changing a zillion files for no real benefit
     "react/jsx-one-expression-per-line": "off",
     "quotes": "off",
```
Comment (HaGuesto) [inline .eslintrc:15]: we agreed on double quotes

File: .eslintrc [hunk 2]
```diff
@@ -9,12 +9,32 @@
   ],
   "plugins": ["jsx-a11y"],
```
Comment (HaGuesto) [inline .eslintrc:10]: move to react folder

File: .eslintrc [hunk 3]
```diff
@@ -9,12 +9,32 @@
...
   "rules": {
+
     // driving me nuts and would have involved changing a zillion files for no real benefit
     "react/jsx-one-expression-per-line": "off",
     "quotes": "off",
     "semi": "off",
     "camelcase":"off",
     "no-unused-expressions": "off",
```
Comment (HaGuesto) [inline .eslintrc:18]: what is the difference to line 34 (vars)

File: .prettierrc [hunk 4]
```diff
@@ -1,6 +0,0 @@
-{
```
Comment (HaGuesto) [inline .prettierrc:1]: is this file needed for backend?

File: .eslintrc [hunk 5]
```diff
@@ -9,12 +9,32 @@
...
+    // ],
+    "no-loop-func":"error",
+    "indent": ["error", 4],
+    "no-tabs": 0,
+    "@typescript-eslint/no-unused-vars": "off",
+    "no-unused-vars": "warn",
+    "no-console": "error",
+    "no-debugger": 1,
     // matter of preference
     "import/prefer-default-export": "off",
```
Comment (HaGuesto) [inline .eslintrc:39]: what is this for?

**Review by HaGuesto — COMMENTED** [review]
naming convention missing

File: .eslintrc (hunk 3 above)
Comment (aerinsol) [reply .eslintrc:18]: Can we also remove the "Driving me nuts" line, it's an old comment from @DurkoMatko

File: react/README.md [hunk 6]
```diff
@@ -72,4 +83,5 @@ To eliminate repetitive code, a custom renderer was built in `react/src/utils/te
 * Ideally only one component per file
 * Files and folders which export a component/view are written UpperCamelCase, with the same name as the actual exported component/view
 * Other files (like types.ts, helpers.ts etc) and folders (like providers, utils etc) are written in lowerCamelCase
-
+* Config constants should be UPPERCASE_SNAKES
+* GraphQL queries, mutation and subscription  **string**  have the format UPPERCASE_SNAKES_<QUERY|MUTATION|SUBSCRIPTION>
```
Comment (aerinsol) [inline react/README.md:87]: Aren't all existing queries mutations etc. currently lowercase_snakes @pylipp ?

Comment (aerinsol) [reply react/README.md:87]: althought maybe it doesn't matter because FE/BE is separate.

File: .eslintrc (hunk 5 above)
Comment (vahidbazzaz) [reply .eslintrc:37]: @HaGuesto When there is only a single export from a module, prefer using default export over named export.

File: react/README.md (hunk 6 above)
Comment (pylipp) [reply react/README.md:87]: The note here refers to an actual query string, not the definition in the schema (GraphQL query/mutation definitions are camelCase).

**Comment by pylipp** [comment]
@vahidbazzaz can you document what commands should be run for linting and formatting? Is it `yarn lint` and `yarn format:write`?

File: react/README.md [hunk 7]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
-* GraphQL queries, mutation and subscription  **string**  have the format UPPERCASE_SNAKES_<QUERY|MUTATION|SUBSCRIPTION>
+-   Views of react-router paths go into the views folder
+-   Each view can have it's own folder - which in return can have a local components folder
+-   Following an "As local as possible, as global as needed" approach: components get only moved into a more global/outer folder if they are used on that level
+-   No index.ts files, besides the entry file for the app
+-   Ideally only one component per file
+-   Files and folders which export a component/view are written UpperCamelCase, with the same name as the actual exported component/view
+-   Other files (like types.ts, helpers.ts etc) and folders (like providers, utils etc) are written in lowerCamelCase
+-   Config constants should be UPPERCASE_SNAKES
+-   GraphQL queries, mutation and subscription **string** have the format UPPERCASE*SNAKES*<QUERY|MUTATION|SUBSCRIPTION>
```
Comment (pylipp) [inline react/README.md:110]: Should be `UPPERCASE_SNAKES_`, otherwise it's rendered italic.

File: react/README.md [hunk 8]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
+
+```bash
+
+/react
+├── /src
+│   ├── assets
+│   ├── components
+│   │   ├── <NameOfComponent>
+│   │   |    ├── <NameOfComponent>Container.tsx
+│   |   |    └── <NameOfComponent>.stoties.tsx #storybook component definition
```
Comment (pylipp) [inline react/README.md:138]: typo: stories :)

File: react/README.md [hunk 9]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
+│   |   └── Layout.tsx # main layout
+│   ├── mocks
+│   ├── providers #context providers
+│   ├── types #typescript definitions
+│   ├── views
+│   │   └── <NameOfComponent>
+│   │       ├──components
+│   |       └── <NameOfSubComponent>.tsx
+│   └── utils
+│           ├── bese-types.ts
```
Comment (pylipp) [inline react/README.md:148]: typo: base-types

File: react/README.md [hunk 10]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
+├── node_modules
+├── .storybook
+├── public
+├── test
+├── App.tsx
+├── index.tsx
+├── logo.svg
+├── serviceWorker.js
+├── setupTests.js
+├── craco.confog.js
```
Comment (pylipp) [inline react/README.md:162]: typo: config

File: react/README.md [hunk 11]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
+│   │   └── <NameOfComponent>
+│   │       ├──components
+│   |       └── <NameOfSubComponent>.tsx
+│   └── utils
+│           ├── bese-types.ts
+│           ├── helpers.ts
+│           ├── queries.ts
+│           ├── test-utils.ts
+│           └── hooks.ts
+├── node_modules
```
Comment (pylipp) [inline react/README.md:153]: Not sure if I'd list the entire content here, it quickly becomes out-of-sync. If people want to have an overview, they should `ls` or see their file-explorer. It makes sense for the `src/` part though.

File: react/README.md [hunk 12]
```diff
@@ -73,15 +87,90 @@ For integration tests, we mock the Apollo client with a `MockedProvider` compone
...
+├── craco.confog.js
+├── codegen.yml
+├── Dockerfile
+├── README.md
+├── package.json
+├── yarn.lock
+├── tsconfig.json
+├── .prettierignore
+├── .dockerignore
+├── .prettierignore
```
Comment (pylipp) [inline react/README.md:171]: duplicated

**Review by pylipp — CHANGES_REQUESTED** [review]
Thanks for extending the README. In another PR we should look at how to integrate the FE linters and formatters as pre-commit hook and in CircleCI. Now that the configs are set up I'd also feel comfortable to do so @HaGuesto @aerinsol

**Review by pylipp — APPROVED** [review]

File: react/README.md (hunk 11 above)
Comment (vahidbazzaz) [reply react/README.md:153]: @pylipp I thought that since most of these files are used for configuration and are unlikely to change, it could be helpful to show them in the docs, but I totally agree with you that they can be easily verified via `ls` or file-explorer.

**Review by aerinsol — APPROVED** [review]

**Comment by pylipp** [comment]
@vahidbazzaz do you realize that `yarn lint` is not effective yet due to `*.tsx` in `.eslintignore`? Anyways that shouldn't be a task for this PR. I created a [follow-up](https://trello.com/c/qnk6KSZa/973-20-apply-linting-formatting-to-relevant-code-in-fe). If you merge master one last time (the dependabot update merged by @aerinsol created a conflict), we can merge this PR.
